### PR TITLE
Access Control and Dataset Lock Documentation

### PR DESCRIPTION
**Pull request by Rosspet**
PR for documentation for access controls and dataset lock.
 Already have had a chat and noted some things re dataset lock but peter yet to read through access controls:

TODO
- upload tab disabled to prevent to dataset data or metadata, and API's will respond with error.
- lock symbol
- screen shots
- trace of lock history
- what to put as a reason. 
- "completed initial round of modelling. outputs settled."

noting also that screenshots are yet to be added to either sections.

Requesting Peter as he is best placed to comment on the accuracy of the documentation as it comments/explains the respective features.

### CODE REVIEW

File: information-system/data-store/access-control.md [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+  </summary>
+{: .text-delta }
+* TOC
+{:toc}
+____
+</details>
+
+# Access Control Overview
+
+This document will discuss the types of access controls available and then reveal the methods to apply these access controls through the Datastore User Interface.
```
Comment (PeterBaker0) [inline information-system/data-store/access-control.md]: Propose "Data Store user interface" instead of "Datastore User Interface"

Comment (PeterBaker0) [reply information-system/data-store/access-control.md]: Propose "discuss" rather than "reveal"

File: information-system/data-store/access-control.md [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+* TOC
+{:toc}
+____
+</details>
+
+# Access Control Overview
+
+This document will discuss the types of access controls available and then reveal the methods to apply these access controls through the Datastore User Interface. 
+
+## What access controls are available for datastore datasets?
```
Comment (PeterBaker0) [inline information-system/data-store/access-control.md]: Change "datastore datasets" to just "datasets"

File: information-system/data-store/access-control.md [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+There are three access rights to consider: 
+
+1. Metadata access
+2. Data access
+3. Admin access
+
+
+## Metadata access
+
+Facilitates control of general viewing and metadata updating.
```
Comment (PeterBaker0) [inline information-system/data-store/access-control.md]: Simplify language here - propose "Metadata access determines if a user can discover, view and update information about a dataset."

File: information-system/data-store/access-control.md [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+### Metadata Read
+
+- If enabled, users with this right will be able view the dataset's metadata and find the dataset in the datastore "search" and "list" tools.
```
Comment (PeterBaker0) [inline information-system/data-store/access-control.md]: Would propose 
> If granted, users will be able to find the dataset in the Data Store and view the dataset's metadata

File: information-system/data-store/access-control.md [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+
+## Metadata access
+
+Facilitates control of general viewing and metadata updating.
+
+### Metadata Read
+
+- If enabled, users with this right will be able view the dataset's metadata and find the dataset in the datastore "search" and "list" tools. 
+- If disabled, users will not be able to find the dataset in the datastore let alone view the metadata.
```
Comment (PeterBaker0) [inline information-system/data-store/access-control.md]: Proposing
> If not granted, users will not be able to find the dataset in the Data Store nor view the dataset's metadata.

File: information-system/data-store/access-control.md [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+
+
+### Metadata Write
+
+- If enabled, users can update metadata fields and will be granted metadata read access regardless of the setting for metadata read.
```
Comment (PeterBaker0) [inline information-system/data-store/access-control.md]: Change "enabled" to "granted" and disabled to "not granted" for all entries

Comment (PeterBaker0) [reply information-system/data-store/access-control.md]: Change "metadata fields" to "the dataset metadata"

File: information-system/data-store/access-control.md [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+Facilitates control of general viewing and metadata updating.
+
+### Metadata Read
+
+- If enabled, users with this right will be able view the dataset's metadata and find the dataset in the datastore "search" and "list" tools. 
+- If disabled, users will not be able to find the dataset in the datastore let alone view the metadata.
+
+
+### Metadata Write
+
```
Comment (PeterBaker0) [inline information-system/data-store/access-control.md:45]: Would propose a "Automatically grants" section below each access level to more clearly describe the inherited access e.g. 

### Metadata Write

#### Also grants

- Metadata read

File: information-system/data-store/access-control.md [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+- If enabled, users can update metadata fields and will be granted metadata read access regardless of the setting for metadata read.
+- If disabled, general users will not be able to update the metadata.
```
Comment (PeterBaker0) [inline information-system/data-store/access-control.md]: Don't include "general" here - this is not specific to general vs groups yet

File: information-system/data-store/access-control.md [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+
+## Data access
+
+Facilitates control over the upload/download of data to/from the storage location associated with the dataset.
```
Comment (PeterBaker0) [inline information-system/data-store/access-control.md]: Simplify: 

> Data access determines if a user can download, upload and modify the files within a dataset.

File: information-system/data-store/access-control.md [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+
+### Data read
+
+- If enabled, the download panel will become visible to users and they will be able to download data from the storage location.
```
Comment (PeterBaker0) [inline information-system/data-store/access-control.md]: Change "data from the storage location" to "the dataset's files"

File: information-system/data-store/access-control.md [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+- If disabled, general users will not be able to update the metadata.
+
+## Data access
+
+Facilitates control over the upload/download of data to/from the storage location associated with the dataset.
+
+### Data read
+
+- If enabled, the download panel will become visible to users and they will be able to download data from the storage location. 
+- If disabled, the download panel will not be visible and users will not be able to download data from the storage location.
```
Comment (PeterBaker0) [inline information-system/data-store/access-control.md]: Same thing "dataset's files"

File: information-system/data-store/access-control.md [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+
+### Data write
+
+- If enabled, users will be able to upload/overwrite and download to/from the storage location as both the upload and download panels will be made available to general users.
```
Comment (PeterBaker0) [inline information-system/data-store/access-control.md]: Don't use "general" here. 

This is worded confusingly. 

Maybe 

> If granted, users will have complete control over the dataset's files - this includes upload, modification and deletion.

File: information-system/data-store/access-control.md [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+- If disabled, users will not be able to upload/overwrite data in the storage location.
```
Comment (PeterBaker0) [inline information-system/data-store/access-control.md]: avoid forward slashes - it makes things more difficult to read.

Maybe 

> If not granted, users will not be able to modify the dataset's files.

File: information-system/data-store/access-control.md [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+
+## Admin access
+
+- If enabled, provides administrative control over the dataset. That is, the ability to lock and unlock the dataset in addition to all of the above privileges. For more information on locking and unlocking, see here.
```
Comment (PeterBaker0) [inline information-system/data-store/access-control.md]: "See here" should now be linked?

Comment (PeterBaker0) [reply information-system/data-store/access-control.md]: I would like "lock and unlock" directly to the locking a dataset page rather than explicitly calling it out. 

Also worth mentioning that admin control provides access control i.e. 

## Admin access

If granted, provides full control over a dataset. Administrators of a dataset are granted all of the above permissions, and can additionally: 

- reconfigure access (see below)
- apply or remove _dataset locks_ (linked)

File: information-system/data-store/access-control.md [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+
+# Applying Access Controls
+
+The aforementioned controls can be applied for individual datasets by those who have administration access to the dataset. By default, this is the owner of the dataset and Information System Admins.
```
Comment (PeterBaker0) [inline information-system/data-store/access-control.md]: "administrative" not administration

Comment (PeterBaker0) [inline information-system/data-store/access-control.md]: "Information System Admins" could just be "Data Store administrators".

File: information-system/data-store/access-control.md [hunk 16]
```diff
@@ -0,0 +1,101 @@
...
+
+Each dataset can have a rule for its accessibility to (the group of all) general datastore users, and (optionally) extra rules for multiple custom groups of users. The following sections will discuss these.
```
Comment (PeterBaker0) [inline information-system/data-store/access-control.md]: Propose 

> Each dataset specifies an access configuration. This configuration defines a set of access grants for a selection of groups. All datasets must specify at least one set of access grants for the default "General Users" group, as defined below. Users can configure extra access grants for custom user groups as discussed below.

File: information-system/data-store/access-control.md [hunk 17]
```diff
@@ -0,0 +1,101 @@
...
+
+- If enabled, provides administrative control over the dataset. That is, the ability to lock and unlock the dataset in addition to all of the above privileges. For more information on locking and unlocking, see here.
+
+# Applying Access Controls
+
+The aforementioned controls can be applied for individual datasets by those who have administration access to the dataset. By default, this is the owner of the dataset and Information System Admins.
+
+Each dataset can have a rule for its accessibility to (the group of all) general datastore users, and (optionally) extra rules for multiple custom groups of users. The following sections will discuss these.  
+
+## General user access to dataset
```
Comment (PeterBaker0) [inline information-system/data-store/access-control.md]: Maybe just

> The "General User" group

Comment (PeterBaker0) [reply information-system/data-store/access-control.md]: Although the data store settings uses the term "General Access" ... hmm

File: information-system/data-store/access-control.md [hunk 18]
```diff
@@ -0,0 +1,101 @@
...
+
+### Who are general users?
+
+General Datastore users refers to the group of all users who can access the datastore. As such, general user access controls facilitate the general privacy or publicity (to all datastore users) of a dataset. You could limit access controls for general users if you wish to keep your dataset private to a group (perhaps during development processes and then later release it by updating general user access rights).
```
Comment (PeterBaker0) [inline information-system/data-store/access-control.md]: Prefer "Data Store" over datastore consistently

Comment (PeterBaker0) [reply information-system/data-store/access-control.md]: This is a bit wordy

Maybe

> The "General Users" group contains all users of the Data Store. Each dataset must specify a set of access grants for this group. This access could range from full privacy (no access grants), to complete control (data write).

There is an edge case here which is that the limiting factor for a user is still their overall permissions in the data store - not sure where that fits in.

File: information-system/data-store/access-control.md [hunk 19]
```diff
@@ -0,0 +1,101 @@
...
+
+### Where to manage general user access
+
+Upon registering a dataset, the general datastore user access rights can be managed by navigating to the 'settings' tab when viewing the dataset.
+
+In future versions of the IS, the ability to set access levels during the registration process will be implemented. For now, a workaround is to register a dataset with meaningless metadata, limit general access privileges, then meaningfully update the metadata as you wish.
```
Comment (PeterBaker0) [inline information-system/data-store/access-control.md]: This paragraph can be removed IMO.

Comment (PeterBaker0) [reply information-system/data-store/access-control.md]: This definitely needs a screenshot. And maybe a 'step by step' guide i.e. Click the settings tab button (highlighted in screenshot), click expand on General Access (highlighted in screenshot), make a set of changes (highlighted in screenshot) and confirm (screenshot).

File: information-system/data-store/access-control.md [hunk 20]
```diff
@@ -0,0 +1,101 @@
...
+
+## Group access to dataset
+
+The group access feature enables custom permissions for groups (and thus members) to a dataset. This can achieve the privatisation of datasets to only certain users of the information system by *not* enabling general user read/write permissions and creating a group with enabled read/write permissions for its members.
```
Comment (PeterBaker0) [inline information-system/data-store/access-control.md]: Simplify this, propose: 

> If granting access to all users does not provide enough control, a dataset administrator can configure access grants on a group-by-group basis. Configuring access for a group grants this access to all registered members of the group.

File: information-system/data-store/access-control.md [hunk 21]
```diff
@@ -0,0 +1,101 @@
...
+
+### Creating a group
+
+Group creation is currently performed by Information System Administrators. Please get in touch and we can help facilitate this for you.
```
Comment (PeterBaker0) [inline information-system/data-store/access-control.md:177]: Jonathan to add section here.

File: information-system/data-store/access-control.md [hunk 22]
```diff
@@ -0,0 +1,101 @@
...
+
+### Adding group access to datasets
+
+Groups access to datasets can be manually managed by navigating to the dataset's settings tab, then Access Control->Groups. Here, group permission can be added to the dataset. Group access levels for this dataset can also be edited here by selecting the group.
```
Comment (PeterBaker0) [inline information-system/data-store/access-control.md]: Remove "manually"

File: information-system/data-store/access-control.md [hunk 23]
```diff
@@ -0,0 +1,101 @@
...
+
+Group access can be automatically applied with default values upon the creation of a new dataset. This occurs if 'Default data store access configuration' is enabled in the group creation process by the IS administrator. The result is that if a user belonging to that group creates a new dataset, then it will automatically be made available to the group according to the specified default access levels. Access levels can then be updated from default by navigating to the group in the dataset as mentioned in the above paragraph.
```
Comment (PeterBaker0) [inline information-system/data-store/access-control.md]: Don't include this - I don't think we have the intention of configuring groups like this until the feature is more mature.

File: information-system/data-store/access-control.md [hunk 24]
```diff
@@ -0,0 +1,101 @@
...
+
+A dataset can be made available to more than one group with differing accessibility levels (see note below). For example, you may wish to have a group of dataset contributors which will require write access, and another group for dataset viewing only, meanwhile the more general system users may not have access at all. 
+
+**Note**: If a user is accessing a dataset which has access rules for multiple groups of which the user is all apart of, the user will be given the more permissive access rights. This is because enabling an access right is seen as a "granting" action, meanwhile not enabling an access right is *not* seen as 'denying' access. Therefore, roles can only grant access, not deny.
```
Comment (PeterBaker0) [inline information-system/data-store/access-control.md]: This can be made a bit clearer:

**Note**: When evaluating a user's access to a dataset, an optimistic approach is taken. For a given action, the following access checks are applied in this order:

1. Does the user have the correct data store access role for this action (see link to access roles doco)? If not, the **action is denied.** 
2. Does the user own this dataset? If so, the **action is approved**.
3. Does the "General Access" configuration for the dataset grant this action? If so, **action is approved**.
4. Does the user belong to **any** group which **grants** this action? If so, **action is approved**.
5. Otherwise, the **action is denied**.

File: information-system/data-store/access-control.md (hunk 22 above)
Comment (PeterBaker0) [inline information-system/data-store/access-control.md]: This whole section needs more detail/screenshots. Will re-review once updated.

**Review by PeterBaker0 — CHANGES_REQUESTED** [review]
Looks good - some wording is a bit technical/unclear - so have suggested a lot of changes. 

The overall structure is good and the content is comprehensive. 

Need more "hand holding" here - screenshots, more detailed instructions. 

I would also appreciate being told some "common uses" or "common configurations" to get a bit of a picture as a user of how we designed the system to be used. 

E.g. "If you would like all users to be able to view and download your dataset, use this configuration -> screenshot". "If you would like all users to be able to view and download your dataset, but only your team's group to be able to edit the dataset, use this configuration -> screenshot".

File: information-system/data-store/access-control.md (hunk 23 above)
Comment (Rosspet) [reply information-system/data-store/access-control.md]: removed line 97.

File: information-system/data-store/access-control.md (hunk 24 above)
Comment (Rosspet) [reply information-system/data-store/access-control.md]: Are you suggesting replacing the whole **note** with these steps? 

I disagree that providing 5 steps clearer than the first sentence in the note. I don't think users need to know the order if it is functionally the same as them gaining access in line with their most permissive rights.

File: information-system/data-store/access-control.md (hunk 1 above)
Comment (PeterBaker0) [reply information-system/data-store/access-control.md]: Resolved

File: information-system/data-store/access-control.md (hunk 2 above)
Comment (PeterBaker0) [reply information-system/data-store/access-control.md]: Resolved

File: information-system/data-store/access-control.md (hunk 3 above)
Comment (PeterBaker0) [reply information-system/data-store/access-control.md]: Resolved

File: information-system/data-store/access-control.md (hunk 4 above)
Comment (PeterBaker0) [reply information-system/data-store/access-control.md]: Resolved

File: information-system/data-store/access-control.md (hunk 5 above)
Comment (PeterBaker0) [reply information-system/data-store/access-control.md]: Resolved

File: information-system/data-store/access-control.md (hunk 21 above)
Comment (PeterBaker0) [reply information-system/data-store/access-control.md:177]: Out of scope for this PR - still okay to approve assuming other stuff resolved

File: information-system/data-store/access-control.md (hunk 19 above)
Comment (PeterBaker0) [reply information-system/data-store/access-control.md]: This paragraph is still here -was that intentional?

File: information-system/data-store/access-control.md (hunk 22 above)
Comment (PeterBaker0) [reply information-system/data-store/access-control.md]: Nice - resolved

File: information-system/data-store/access-control.md (hunk 23 above)
Comment (PeterBaker0) [reply information-system/data-store/access-control.md]: Resolved

File: information-system/data-store/access-control.md (hunk 24 above)
Comment (PeterBaker0) [reply information-system/data-store/access-control.md]: Maybe you're right - I think my description above is better for our dev documentation. 

I think you're wording could be simplified in that case, maybe just

**Note** If a user is accessing a dataset which has access rules for multiple groups of which the user is a part of, the user will be granted the most permissive combination of those access rights.

File: information-system/data-store/access-control.md (hunk 22 above)
Comment (PeterBaker0) [reply information-system/data-store/access-control.md]: Looks okay - wording is possibly a little verbose but okay for first merge. Will be easier to see and review small bits as we go through a final read.

Comment (PeterBaker0) [reply information-system/data-store/access-control.md]: Screenshots are good - thx

**Review by PeterBaker0 — CHANGES_REQUESTED** [review]
Missing some of the unexpanded comments. More review still to come once next round done.

File: information-system/data-store/access-control.md (hunk 7 above)
Comment (PeterBaker0) [reply information-system/data-store/access-control.md:45]: Resolved

File: information-system/data-store/access-control.md (hunk 6 above)
Comment (PeterBaker0) [reply information-system/data-store/access-control.md]: Resolved

File: information-system/data-store/access-control.md (hunk 8 above)
Comment (PeterBaker0) [reply information-system/data-store/access-control.md]: Resolved

File: information-system/data-store/access-control.md (hunk 9 above)
Comment (PeterBaker0) [reply information-system/data-store/access-control.md]: Resolved

File: information-system/data-store/access-control.md (hunk 11 above)
Comment (PeterBaker0) [reply information-system/data-store/access-control.md]: Resolved

File: information-system/data-store/access-control.md (hunk 12 above)
Comment (PeterBaker0) [reply information-system/data-store/access-control.md]: Resolved

File: information-system/data-store/access-control.md (hunk 13 above)
Comment (PeterBaker0) [reply information-system/data-store/access-control.md]: Resolved

File: information-system/data-store/access-control.md (hunk 14 above)
Comment (PeterBaker0) [reply information-system/data-store/access-control.md]: Resolved

File: information-system/data-store/access-control.md (hunk 15 above)
Comment (PeterBaker0) [reply information-system/data-store/access-control.md]: Resolved

Comment (PeterBaker0) [reply information-system/data-store/access-control.md]: Resolved

File: information-system/data-store/access-control.md (hunk 16 above)
Comment (PeterBaker0) [reply information-system/data-store/access-control.md]: Resolved

File: information-system/data-store/access-control.md (hunk 17 above)
Comment (PeterBaker0) [reply information-system/data-store/access-control.md]: Resolved

File: information-system/data-store/access-control.md (hunk 18 above)
Comment (PeterBaker0) [reply information-system/data-store/access-control.md]: Resolved

File: information-system/data-store/access-control.md (hunk 21 above)
Comment (PeterBaker0) [reply information-system/data-store/access-control.md:177]: @jyucsiro  just noting that this will require your input here regarding the proper procedure to request a group.

File: information-system/data-store/access-control.md (hunk 24 above)
Comment (PeterBaker0) [reply information-system/data-store/access-control.md]: Went with above

Comment (PeterBaker0) [reply information-system/data-store/access-control.md]: Resolved

File: information-system/data-store/access-control.md (hunk 22 above)
Comment (PeterBaker0) [reply information-system/data-store/access-control.md]: Added some more detailed steps/screenshots.
